### ISSUE TITLE
base64 decode detection with X509::_extractBER not always working correctly

### ISSUE THREAD

**Issue by PadaKwaak** [phpseclib/phpseclib#854]
I'm having an issue when calling X509::loadCSR with a DER encoded CSR, where it incorrectly detects/strips PEM encoding. Would it be possible to add an argument to indicate that the input is DER encoded and then skip the PEM format detection/parsing?

Here is my CSR (which I've PEM encoded just to submit it here) that wasn't parsed correctly:

```
-----BEGIN CERTIFICATE REQUEST-----
MIICdzCCAV8CAQEwDDEKMAgGA1UEAwwBeDCCASIwDQYJKoZIhvcNAQEBBQADggEP
ADCCAQoCggEBALtcrFDD2AHe3x2bR00wPDsPH6FJLxr5uc1ybb+ldDB5xNVImC8P
LU6VXDZ5z68KjSovs1q0OWJWfCjlAuGLzqO35s86LI1CFuTFdkScVHMwh8zUVFoP
pG7/9rKaNxCgaHs4evxjxQP2+Ny7tBqPLb/KV0exm6Twocf963jC/Tyn57G5erRf
zpFrfK7DozhxY7znumJ4FuSn0TVkD6PPwZFn9VoTjv2ZoJmacGK+0r5yNKG799F5
K8EgDrOCfbzCZjX6GJctyn2SNPTeBuXS9piH21FGnJAryv80zG+zUqFdEyoLUGJt
4Vy6+tDP9cW68fiwTZS1Oc1VeFdL1G/CrjkCAwEAAaAmMCQGCSqGSIb3DQEJDjEX
MBUwEwYKKwYBBAGCqlsBCQQFMAOCAQEwDQYJKoZIhvcNAQELBQADggEBAF4XOd+1
jkJOYRInNpHfhzSD/ktDY50gpLPuDvl4f/ZBlKrb1eDYQG5F3bnYzoZWHN4n+6Zs
CkljXs5ZPUZ5LuVpASumoG/aHXGz8c8NC3asJ1V73ljEPAfIXwqoIUoaP9jLL+Ee
zy/ZCi2NKWVo2D7ocnn79oblAem9ksSeQl4z3Gvhuug6MsMqn96NU/ZY/vjYzAjb
MAvJIVRY0rbCxbFa0K+XNJtF7GLyBxyPNFWCvADhvm9C4uPmoypYg7MY6EewJInN
xzMH7I4xDLjNu0VBa6lAxTvflp0joQHKlTYX0SDIKPbQivjZMuObPuxDtkVZ0rQl
AjmgMowaN5otTXM=
-----END CERTIFICATE REQUEST-----
```

At the moment I have it working by simply doing a base64_encode operation before parsing my DER encoded CSR as an argument.

**Comment by terrafrost** [comment]
#855 improves the base64-encoded detection. Adding a second parameter isn't a bad idea but it'll take more time. Maybe I could make it so you can pass constants to it. eg. X509::FORMAT_AUTO_DETECT, X509::FORMAT_PEM, X509::FORMAT_DER. But I can do that later. For the time being there's #855.

**Comment by PadaKwaak** [comment]
Thanks for the fast feedback & work done on this project!

I'll keep an eye out for progress on #855 and do as you suggested with the 3 different constants.

**Comment by terrafrost** [comment]
I've implemented this in RSA.php as well since it pretty much has the exact same _extractBER() function.

For X509 adding a second parameter to the various load methods works. For RSA it doesn't. I'm not going to add custom parameters to that classes loadKey() method that only mean anything for select key formats.

I suppose I could add PRIVATE_FORMAT_PKCS1_PEM (for base64-encoded stuff) and PRIVATE_FORMAT_PKCS1_DER (for non-base64-encoded stuff) and have PRIVATE_FORMAT_PKCS1 auto detect (which is what it currently does) but tbh I just don't want to. I don't want to have fifty billion different formats that are all just super slight variations of one another. Especially when auto detection works 99% of the time (even in X509.php and even more so after this commit).

For the rsa-plugins branch (which'll be merged into master and only master; see #837) I could do something like `PKCS1::setMode(PKCS1::PEM)` or something. That's one of the advantages of the plugin approach - you can have per format features.

The use of constants in X509.php means that if the load*() functions were to start using plugins like RSA.php that that'd be a BC break but I don't think that's a huge issue as I've never heard X509 certs in any format other than DER / PEM. There are the XML Encoding Rules but I've never once seen those in use in the wild.

**Comment by terrafrost** [comment]
So for the master branch you can now do `PKCS1::requirePEM()`, `PKCS1::requireDER()` or `PKCS1::requireAny()`.
